File: 15-3sum/3sum.ts
function threeSum(nums: number[]): number[][] {
    nums.sort((a, b) => a - b); 

    const results: number[][] = [];
    for (let i = 0; i < nums.length - 2; i++) {
        // Skip duplicate elements to avoid duplicate triplets
        if (i > 0 && nums[i] === nums[i - 1]) {
            continue;
        }
        
        let l = i + 1;
        let r = nums.length - 1;
        
        while (l < r) {
            const total = nums[i] + nums[l] + nums[r];
            if (total === 0) {
                results.push([nums[i], nums[l], nums[r]]);
                
                // skipping duplicates , same as above
                while (l < r && nums[l] === nums[l + 1]) {
                    l++;
                }
                while (l < r && nums[r] === nums[r - 1]) {
                    r--;
                }
                l++;
                r--;
            } else if (total < 0) {
                l++;
            } else {
                r--;
            }
        }
    }
    return results;
}

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        res = []
        for i,num in enumerate(nums):
            if i > 0  and nums[i] == nums[i-1]:
                continue
            l,r =i+1,len(nums)-1
            while l < r:
                ans = nums[i] + nums[l] + nums[r]
                if ans == 0:
                    res.append([nums[i], nums[l], nums[r]])
                    l +=1 
                    while l < r and nums[l] == nums[l-1]:
                        l +=1
                    r -= 1
                    while l < r and nums[r] == nums[r+1]:
                        r -=1
                elif ans < 0:
                    l +=1
                else:
                    r -=1
        return res
                    
                 